aten/src/ATen/native/mkldnn/RNN.cpp

```diff
@@ -0,0 +1,583 @@
+        grad_cy_, rnn.dst_iter_desc(get_mkldnn_dtype(grad_cy_.scalar_type())));
+  } else {
+    diff_y = get_mkldnn_tensor(
+        grad_output, rnn.dst_layer_desc(ideep::tensor::data_type::f32));
+    diff_hy = get_mkldnn_tensor(
+        grad_hy, rnn.dst_iter_desc(ideep::tensor::data_type::f32));
+    diff_cy = get_mkldnn_tensor(
+        grad_cy, rnn.dst_iter_desc(ideep::tensor::data_type::f32));
+  }
+
+  auto forward_hint = ideep::lstm_forward_training::prepare(x, hx, cx, w1, w2, b, y, hy, cy, reverse);
+  ideep::tensor mkldnn_workspace;
+  mkldnn_workspace.init(
+      forward_hint.workspace_desc(), workspace.template data_ptr<uint8_t>());
+  ideep::lstm_backward::compute(forward_hint, x, hx, cx, w1, w2, b, y, hy, cy, diff_y, diff_hy, diff_cy, mkldnn_workspace, diff_x, diff_hx, diff_cx, diff_w1, diff_w2, diff_b, reverse);
+  auto diff_b2_ = at::clone(diff_b_);
+  return std::make_tuple(diff_x_, diff_w1_, diff_w2_, diff_b_, diff_b2_, diff_hx_, diff_cx_);
+}
+
+// MKLDNN RNN integration notes:
+// I. Memory Formats
+//   a. mkldnn will use plain formats for input, hx/cx, output, hy/cy
+//      and possibly use blocked formats for weights depending shape info.
+//   b. All mkldnn memorys are created (in plain format) as views on ATen tensor,
+//      the weight reorder(if any) is handed automatically inside ideep (mkldnn bridge)
+//
+// II. MKLDNN Primitive Mapping
+//   a. mkldnn rnn primitive doesn't support training with dropout or padded input sequence.
+//   b. here break a single RNN module into { num_layers * num_directions } mkldnn rnn primitives
+//      for future need to cover these feature gaps.
+//
+//TODO: a. training with dropout
+//   b. padded sequence input support
+//
+
+std::tuple<Tensor, Tensor, Tensor> mkldnn_rnn(
+    const Tensor& input_, TensorList weight, int64_t weight_stride0,
+    const Tensor& hx_, const Tensor& cx_,
+    int64_t mode, int64_t hidden_size,
+    int64_t num_layers, bool has_biases, bool batch_first, double dropout_p,
+    bool train, bool bidirectional, IntArrayRef batch_sizes) {
+  TORCH_CHECK(batch_sizes.size() == 0, "mkldnn_rnn doesn't support packed input");
+  if (static_cast<ideep::rnn_kind>(mode) != ideep::rnn_kind::LSTM) {
+    TORCH_CHECK(!cx_.defined(), "mkldnn_rnn: illegal defined cx for non-LSTM RNN");
+  }
+
+  auto input = input_;
+  if (batch_first) {
+    input = input.transpose(0, 1);
+  }
+  input = input.contiguous();
+
+  auto hx = hx_.contiguous();
+  auto cx = cx_.contiguous();
+
+  MatrixRef<Tensor> weights{weight, static_cast<size_t>(weight_stride0)};
+
+  auto num_directions = bidirectional ? 2 : 1;
+  auto layer_input = input;
+  std::vector<at::Tensor> layer_output(num_directions);
+  std::vector<at::Tensor> layer_hy(num_layers * num_directions);
+  std::vector<at::Tensor> layer_cy(num_layers * num_directions);
+  for (int64_t layer = 0; layer < num_layers; layer++) {
+    for (int64_t direction = 0; direction < num_directions; direction++) {
```

Comment:
Please use `irange` to express that those values are not mutated inside the loop
```suggestion
  for (const auto layer: c10::irange(num_layers)) {
    for (const auto direction: c10::irange(num_directions)) {
```